### PR TITLE
[SPARK-11848][SQL] Support EXPLAIN in DataSet APIs

### PR DESCRIPTION
**Pull request by gatorsmile**
When debugging DataSet API, I always need to print the logical and physical plans. 

I am wondering if we should provide a simple API for EXPLAIN?

### CODE REVIEW

**Comment by gatorsmile** [comment]
Do you think this makes sense? @rxin @marmbrus @cloud-fan

**Comment by gatorsmile** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -580,6 +580,18 @@ class Dataset[T] private[sql](
...
+   * Prints the plans (logical and physical) to the console for debugging purposes.
+   * @since 1.6.0
+   */
+  def explain(extended: Boolean): Unit = toDF().explain(extended)
+
+  /**
+   * Only prints the physical plan to the console for debugging purposes.
+   * @since 1.6.0
+   */
+  def explain(): Unit = explain(extended = false)
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: How about putting this in `Queryable`?

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: A good idea!

**Comment by gatorsmile** [comment]
@marmbrus I moved the functions `explain` to `Queryable`. Please help me check if the implementation is OK. Thank you!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/Queryable.scala [hunk 2]
```diff
@@ -34,4 +36,23 @@ private[sql] trait Queryable {
...
     }
   }
+
+  /**
+   * Prints the plans (logical and physical) to the console for debugging purposes.
+   * @since 1.3.0
+   */
+  def explain(extended: Boolean): Unit = {
+    val explain = ExplainCommand(queryExecution.logical, extended = extended)
+    DataFrame(sqlContext, explain).queryExecution.executedPlan.executeCollect().foreach {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Queryable.scala]: maybe we can simplify this to `sqlContext.executePlan(explain).excutedPlan......`

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/Queryable.scala]: : ) Sure. Thanks!

**Comment by cloud-fan** [comment]
LGTM pending tests.

**Comment by marmbrus** [comment]
Thanks!  Merging to master and 1.6.

**Comment by gatorsmile** [comment]
Thank you for your helps!

**Pull request closed without merging**